TF_Service_dotNet/TouchFree/HandManager.cs

```diff
@@ -182,12 +197,86 @@ public void Update(object sender, FrameEventArgs e)
+                var secondaryRelativeYScreenPosition = secondaryHandIndexTipLocation.Y / screenHeightPX;
+
+                if (lastPrimaryLocation.HasValue && lastSecondaryLocation.HasValue)
+                {
+                    var primaryLocationChange = lastPrimaryLocation.Value - primaryHandIndexTipLocation;
+                    var primaryLocationChangeX = primaryLocationChange.X / screenWidthPX;
+                    var primaryLocationChangeY = primaryLocationChange.Y / screenHeightPX;
+
+                    var secondaryLocationChange = lastSecondaryLocation.Value - secondaryHandIndexTipLocation;
+                    var secondaryLocationChangeX = secondaryLocationChange.X / screenWidthPX;
+                    var secondaryLocationChangeY = secondaryLocationChange.Y / screenHeightPX;
+
+                    PrimaryHandActivity = Math.Clamp(PrimaryHandActivity * 0.9f + Math.Abs(primaryLocationChangeX) + Math.Abs(primaryLocationChangeY), 0f, 1f);
+                    SecondaryHandActivity = Math.Clamp(SecondaryHandActivity * 0.9f + Math.Abs(secondaryLocationChangeX) + Math.Abs(secondaryLocationChangeY), 0f, 1f);
+                }
+
+                lastPrimaryLocation = primaryHandIndexTipLocation;
+                lastSecondaryLocation = secondaryHandIndexTipLocation;
+
+                if (SecondaryHandIsOnScreen(secondaryRelativeXScreenPosition, secondaryRelativeYScreenPosition) &&
+                    (HandActivityInSwapThreshold() ||
+                     PrimaryHandIsOffScreen(primaryRelativeXScreenPosition, primaryRelativeYScreenPosition)))
+                {
+                    Hand oldPrimaryHand = PrimaryHand;
+                    PrimaryHand = SecondaryHand;
+                    SecondaryHand = oldPrimaryHand;
+                    primaryChirality = PrimaryHand.IsLeft ? HandChirality.LEFT : HandChirality.RIGHT;
+                    secondaryChirality = SecondaryHand.IsLeft ? HandChirality.LEFT : HandChirality.RIGHT;
+                }
+            }
+            else
+            {
+                PrimaryHandActivity = PrimaryHandActivityDefault;
+                SecondaryHandActivity = SecondaryHandActivityDefault;
+                lastPrimaryLocation = null;
+                lastSecondaryLocation = null;
+            }
+
             UpdateHandStatus(ref PrimaryHand, leftHand, rightHand, HandType.PRIMARY);
             UpdateHandStatus(ref SecondaryHand, leftHand, rightHand, HandType.SECONDARY);
 
             HandsUpdated?.Invoke(PrimaryHand, SecondaryHand);
         }
 
+        private bool HandActivityInSwapThreshold()
+        {
+            return PrimaryHandActivity < SecondaryHandActivityDefault && SecondaryHandActivity > PrimaryHandActivityDefault;
+        }
+
+        private static bool PrimaryHandIsOffScreen(float _primaryRelativeXScreenPosition, float _primaryRelativeYScreenPosition)
+        {
+            return _primaryRelativeXScreenPosition > 1.4 ||
```

Comment:
If I'm reading this right does this mean that for a hand to be considered "off-screen" it needs to be more than 40% of the screen's size in that dimension outside the screen? So "on-screen" is an area about 180% of the width and 180% of height vs. the _actual_ screen size?

Reply:
Yeah, that's right, we could potentially reduce that but I don't think it would be good to have it the same as the screen size as it could make clicking buttons near the edge of the screen a pain

Reply:
I agree on not making it a 1-1 relationship, I'm just wondering about the 40% either side? Might be good to get some feedback from Chris on how far this might go. I think that with 40% (especially with larger screens) you're still quite likely to have the "other hand holding coffee" issue quite easily?